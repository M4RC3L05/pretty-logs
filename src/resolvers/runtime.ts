import type { SupportedRuntime } from "../types.ts";
import process from "node:process";
import { fstatSync } from "node:fs";
import { inspect } from "node:util";
import { Readable, Writable } from "node:stream";

export const resolveRuntime = (
  runtime: SupportedRuntime,
): {
  isPiping: boolean;
  stdin: ReadableStream;
  stdout: WritableStream;
  waitFirstSigint: () => void;
  inspect: (data: unknown) => string;
} => {
  if (runtime === "bun") {
    return {
      waitFirstSigint: () => process.once("SIGINT", () => {}),
      inspect: (data) =>
        // deno-lint-ignore ban-ts-comment
        // @ts-ignore
        Bun.inspect(data, { colors: true, depth: 100, sorted: true }),
      isPiping: !process.stdin.isTTY &&
        !fstatSync(process.stdin.fd).isFile(),
      // deno-lint-ignore ban-ts-comment
      // @ts-ignore
      stdin: Bun.stdin.stream() as ReadableStream,
      stdout: new WritableStream({
        write: async (chunk) => {
          // deno-lint-ignore ban-ts-comment
          // @ts-ignore
          await Bun.write(Bun.stdout, chunk);
        },
      }),
    };
  }

  if (runtime === "node") {
    return {
      waitFirstSigint: () => process.once("SIGINT", () => {}),
      inspect: (data) => inspect(data, false, 1000, true),
      isPiping: !process.stdin.isTTY &&
        !fstatSync(process.stdin.fd).isFile(),
      stdin: Readable.toWeb(process.stdin) as ReadableStream,
      stdout: Writable.toWeb(process.stdout),
    };
  }

  if (runtime === "deno") {
    return {
      waitFirstSigint: () => {
        Deno.addSignalListener("SIGINT", function noop() {
          Deno.removeSignalListener("SIGINT", noop);
        });
      },
      inspect: (data) =>
        Deno.inspect(data, {
          colors: true,
          compact: false,
          strAbbreviateSize: Number.POSITIVE_INFINITY,
          depth: 1000,
          sorted: true,
          trailingComma: true,
        }),
      isPiping: !Deno.stdin.isTerminal() &&
        !fstatSync(process.stdin.fd).isFile(),
      stdin: Deno.stdin.readable,
      stdout: Deno.stdout.writable,
    };
  }

  throw new Error(`Runtime "${runtime}" not supported.`);
};
