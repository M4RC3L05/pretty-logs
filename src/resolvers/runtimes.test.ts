import { afterEach, describe, it } from "@std/testing/bdd";
import { restore, stub } from "@std/testing/mock";
import { assertSnapshot } from "@std/testing/snapshot";
import { assertEquals, assertInstanceOf, fail } from "@std/assert";
import process from "node:process";
import fs from "node:fs";
import { resolveRuntime } from "./runtime.ts";
import type { SupportedRuntime } from "../types.ts";

describe("resolveRuntime()", () => {
  afterEach(() => {
    restore();
  });

  it("should throw an error if runtime is supported", () => {
    try {
      // deno-lint-ignore no-explicit-any
      resolveRuntime("foo" as any);

      fail("Should have thrown an error");
    } catch (error) {
      assertInstanceOf(error, Error);
      assertEquals(error.message, 'Runtime "foo" not supported.');
    }
  });

  for (const runtime of ["bun", "deno", "node"] as SupportedRuntime[]) {
    it(`should resolve the runtime for ${runtime}`, async (ctx) => {
      if (["bun", "node"].includes(runtime)) {
        stub(process.stdin, "on");
        stub(process.stdout, "on");
        // deno-lint-ignore no-explicit-any
        stub(fs, "fstatSync", () => ({ isFile: () => true }) as any);
      }

      if (runtime === "deno") {
        // deno-lint-ignore no-explicit-any
        stub(Deno, "fstatSync", () => ({ isFile: () => true }) as any);
      }

      const resolvedRuntime = resolveRuntime(runtime);

      await assertSnapshot(ctx, resolvedRuntime);
    });
  }
});
