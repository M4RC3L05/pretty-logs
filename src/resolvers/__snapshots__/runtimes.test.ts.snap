export const snapshot = {};

snapshot[`resolveRuntime() > should resolve the runtime for bun 1`] = `
{
  inspect: [Function: inspect],
  isPiping: false,
  stdin: ReadableStream {
    locked: false,
  },
  stdout: WritableStream {
    locked: false,
  },
  waitFirstSigint: [Function: waitFirstSigint],
}
`;

snapshot[`resolveRuntime() > should resolve the runtime for deno 1`] = `
{
  inspect: [Function: inspect],
  isPiping: false,
  stdin: ReadableStream {
    locked: false,
  },
  stdout: WritableStream {
    locked: false,
  },
  waitFirstSigint: [Function: waitFirstSigint],
}
`;

snapshot[`resolveRuntime() > should resolve the runtime for node 1`] = `
{
  inspect: [Function: inspect],
  isPiping: false,
  stdin: ReadableStream {
    locked: false,
  },
  stdout: WritableStream {
    locked: false,
  },
  waitFirstSigint: [Function: waitFirstSigint],
}
`;
